lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java

```diff
@@ -92,30 +101,77 @@ public void addNewDocumentEvents(Collection<? extends DocumentEvent> documentEve
+            throws Exception {
+        String[] typesToProcess = getSupportedAndEnabledEventTypes();
+        Integer documentEventsBatchSize = config.getDocumentEventsBatchSize();
+
+        if (typesToProcess.length == 0 || documentEventsBatchSize == null ||
+                documentEventsBatchSize == 0) {
+            logger.info("Not retrieving any document events because either there are no enabled " +
+                    "or supported types to process or documentEventBatchSize is 0. Supported " +
+                    "types are {}. Of those, enabled types are {}. " +
+                    "Document event batch size is {}.",
+                    supportedTypes, Arrays.toString(typesToProcess), documentEventsBatchSize);
+            return Collections.emptyList();
+        }
 
-            DocumentEventEntity[] documentEventEntities = lightblue
-                    .data(FindRequests.priorityDocumentEventsForEntitiesUpTo(entities, documentEventBatchSize))
-                    .parseProcessed(DocumentEventEntity[].class);
+        if (maxEvents == 0) {
+            return Collections.emptyList();
+        }
 
-            if (documentEventEntities.length == 0) {
-                return Collections.emptyList();
-            }
+        DocumentEventEntity[] documentEventEntities = lightblue
+                .data(FindRequests.priorityDocumentEventsForTypesUpTo(
+                        typesToProcess, documentEventsBatchSize,
+                        ZonedDateTime.now(clock).minus(processingTimeout)))
+                .parseProcessed(DocumentEventEntity[].class);
+
+        if (documentEventEntities.length == 0) {
+            return Collections.emptyList();
+        }
+
+        try (LockedResources<SharedIdentityEvents> eventLocks =
+                     SharedIdentityEvents.parseAndOptimizeLockableDocumentEventEntities(
+                             maxEvents,
+                             documentEventEntities,
+                             new BulkLightblueRequester(lightblue),
+                             documentEventFactoriesByType,
+                             lockStrategy,
+                             clock)) {
+
+            return persistNewEntitiesAndStatusUpdatesToExisting(eventLocks);
+        }
+    }
 
-            List<DocumentEventEntity> entitiesToUpdate = new ArrayList<>();
-            BulkLightblueRequester requester = new BulkLightblueRequester(lightblue);
+    /**
+     * {@inheritDoc}
+     *
+     * <p>N.B. This implementation currently works by simply checking if the known in memory
+     * timeouts of provided events fall within some threshold. It could be updated to also timestamp
+     * the persisted events again, but to be safe we would probably want to also lock them to do
+     * this update. For now, that complexity is probably not worth it.
+     */
+    @Override
+    public Collection<? extends DocumentEvent> checkExpired(Collection<? extends DocumentEvent> events) {
```

Comment:
> You don't need to keep a locked resource, because you know the entity id.

Right now we don't know the callerId though. I have callerId randomized per call so that multiple thread sharing same instance don't produce duplicate events. The callerId is captured in the LockedResource. An alternative would be to make each thread use their own repository and lock strategy, or require any one node be single threaded. I thought it was less complicated in the end to use timestamps + expiration.